async_simple/coro/ConditionVariable.h

```diff
@@ -0,0 +1,200 @@
+        ;
+    resumeWaiters(awaitings);
+}
+
+template <class Lock>
+inline void ConditionVariable<Lock>::resumeWaiters(
+    ConditionVariableAwaiter<Lock>* awaiters) {
+    for (auto head = awaiters; head; head = head->_next) {
+        head->_continuation.resume();
+    }
+}
+
+template <>
+class ConditionVariableAwaiter<>;
+template <>
+class ConditionVariable<> {
+public:
+    using pointer_type = void*;
+
+public:
+    ConditionVariable() noexcept {}
+    ~ConditionVariable() {}
+
+    ConditionVariable(const ConditionVariable&) = delete;
+    ConditionVariable& operator=(const ConditionVariable&) = delete;
+
+    void notify() noexcept;
+    ConditionVariableAwaiter<> wait() noexcept;
+    void reset() noexcept;
+
+private:
+    void resumeWaiters(ConditionVariableAwaiter<>* awaiters);
+
+private:
+    friend class ConditionVariableAwaiter<>;
+    std::atomic<pointer_type> _awaiters = nullptr;
+};
+
+template <>
+class ConditionVariableAwaiter<> {
+public:
+    using pointer_type = void*;
+
+public:
+    ConditionVariableAwaiter(ConditionVariable<>* cv) noexcept : _cv(cv) {}
+
+    bool await_ready() const noexcept {
+        return static_cast<pointer_type>(_cv) ==
+               _cv->_awaiters.load(std::memory_order_acquire);
+    }
+    bool await_suspend(std::coroutine_handle<> continuation) noexcept {
+        _continuation = continuation;
+        pointer_type awaitings = _cv->_awaiters.load(std::memory_order_acquire);
+        do {
+            if (awaitings == static_cast<pointer_type>(_cv)) {
+                return false;
+            }
+            _next = static_cast<ConditionVariableAwaiter<>*>(awaitings);
+        } while (!_cv->_awaiters.compare_exchange_weak(
+            awaitings, static_cast<pointer_type>(this),
+            std::memory_order_release, std::memory_order_acquire));
+        return true;
+    }
+    void await_resume() const noexcept {}
+
+public:
+    ConditionVariable<>* _cv;
+
+private:
+    friend class ConditionVariable<>;
+    ConditionVariableAwaiter<>* _next = nullptr;
+    std::coroutine_handle<> _continuation;
+};
+
+inline ConditionVariableAwaiter<> ConditionVariable<>::wait() noexcept {
+    return ConditionVariableAwaiter<>{this};
+}
+
+inline void ConditionVariable<>::notify() noexcept {
+    pointer_type self = static_cast<pointer_type>(this);
+    pointer_type awaitings =
+        _awaiters.exchange(self, std::memory_order_acq_rel);
+    if (awaitings != self) {
+        resumeWaiters(static_cast<ConditionVariableAwaiter<>*>(awaitings));
+    }
+}
+
+inline void ConditionVariable<>::reset() noexcept {
+    pointer_type self = static_cast<pointer_type>(this);
+    _awaiters.compare_exchange_strong(self, nullptr, std::memory_order_relaxed);
+}
+
+inline void ConditionVariable<>::resumeWaiters(
+    ConditionVariableAwaiter<>* awaiters) {
+    for (auto head = awaiters; head; head = head->_next) {
+        head->_continuation.resume();
+    }
```

Comment:
```suggestion
    while (awaiters) {
        auto *prev = awaiters;
        awaiters = awaiters->_next;
        prev->_continuation.resume();
    }
```

We shouldn't run `head = head->_next` after `head->_continuation.resume();`. Since after the resumption, the coroutine the `head` refers might be released.

Reply:
apply this fix can pass the asan check？